`meson wrap install` multiple projects

pobrn opened the pull request:
```
Previously, `meson wrap install` took only one argument, so installing
multiple projects needed multiple invocations, which is not too convenient.
So add support for installing multiple projects.
```

xclaesse commented on mesonbuild/wrap/wraptool.py:
```diff
@@ -90,30 +90,52 @@ def search(options: 'argparse.Namespace') -> None:
-    (version, revision) = get_latest_version(name, options.allow_insecure)
-    url = open_wrapdburl(f'https://wrapdb.mesonbuild.com/v2/{name}_{version}-{revision}/{name}.wrap', options.allow_insecure, True)
+        raise WrapException(f'Wrap file for {name} already exists.')
+    (version, revision) = find_latest_version(get_releases(), name)
```
Didn't you want to optionally get version from the name, something like:
```
wanted = name.rsplit('=', maxsplit=1)
if len(wanted) > 1:
  version, revision = wanted.rsplit('-', maxsplit=1)
else:
  version, revision = find_latest_version(get_releases(), name)
```

xclaesse replied:
Ideally it could search for latest that match the wanted version. Something like `foo=2` would find foo `2.1-1`.

xclaesse replied:
> I would like to address that in a separate PR.

Fair enough.

xclaesse commented on mesonbuild/wrap/wraptool.py:
```diff
@@ -90,30 +90,52 @@ def search(options: 'argparse.Namespace') -> None:
+    releases: T.Dict[str, T.Any] = None
+    failed = 0
+
+    def get_releases_cached() -> T.Dict[str, T.Any]:
```
It's usually not worth optimizing the failure case. But it could be worth simply adding `@lru_cache(maxsize=None)` on `get_releases()` function.

xclaesse replied:
I noticed that status() calls get_releases() for each wrap file... caching it there would be a nice bonus.

xclaesse replied:
Made https://github.com/mesonbuild/meson/pull/12500 for that.

xclaesse commented:
> I'm wondering if merely the existence of the wrap file should constitute a fatal error

I don't think that should be an error, but maybe just print a message telling it can be updated with `meson wrap update`.

xclaesse commented:
Looks good. You can also write a release notes snippet, just add a file in `doc/markdown/snippets`. Other than that, LGTM after 1.3 is branched.

bruchar1 commented on mesonbuild/wrap/wraptool.py:
```diff
@@ -99,21 +99,36 @@ def get_latest_version(name: str, allow_insecure: bool) -> T.Tuple[str, str]:
         f.write(url.read())
     print(f'Installed {name} version {version} revision {revision}')
 
+def install(options: 'argparse.Namespace') -> None:
```
```suggestion
def install(options: argparse.Namespace) -> None:
```

bruchar1 replied:
Quotes are not needed because of the `from __future__ import annotations`. I tend to correct them when I saw them on a line I had to edit anyway.

eli-schwartz replied:
*technically* this line was not edited. It was moved downwards and code was refactored around it.

It is just that github's diff algorithm isn't detecting this line as the same. It may be possible to pass different algorithm options to the `git diff` commandline that cause it to detect the identical line better, though it's not super important one way or another.

bruchar1 reviewed (commented):
Your should also update the manual in docs/markdown/Using-wraptool.md

bruchar1 commented on docs/markdown/Using-wraptool.md:
```diff
@@ -64,7 +64,12 @@ the install command.
 
 Now you can issue a `subproject('libjpeg')` in your `meson.build` file
-to use it.
+to use it. You may also specify multiple dependencies to install.
```
You should specify that the feature is new in meson 1.4.0.

bruchar1 reviewed (commented):
Your should also update the manual in docs/markdown/Using-wraptool.md

eli-schwartz commented on mesonbuild/wrap/wraptool.py:
```diff
@@ -89,21 +89,36 @@ def get_latest_version(name: str, allow_insecure: bool) -> T.Tuple[str, str]:
     if os.path.isdir(os.path.join('subprojects', name)):
-        raise SystemExit('Subproject directory for this project already exists.')
+        print(f'Subproject directory for {name} already exists.')
+        return
```
This is an API change. Formerly, if a wrap already existed and the `meson wrap install` failed, we raised SystemExit, now we ignore the issue entirely.

I understand you don't want to raise SystemExit because other wraps should have a chance to be installed too, but in that case we should switch to raising a WrapException.

eli-schwartz commented on mesonbuild/wrap/wraptool.py:
```diff
@@ -89,21 +89,36 @@ def get_latest_version(name: str, allow_insecure: bool) -> T.Tuple[str, str]:
-        raise SystemExit('Wrap file already exists.')
-    (version, revision) = get_latest_version(name, options.allow_insecure)
-    url = open_wrapdburl(f'https://wrapdb.mesonbuild.com/v2/{name}_{version}-{revision}/{name}.wrap', options.allow_insecure, True)
+        print(f'Wrap file for {name} already exists.')
+        return
```
Same here.

panfengli-work commented:
why not merge this pr, it is already approved, it helps usage of `meson wrap install package@version`.

eli-schwartz commented on mesonbuild/wrap/wraptool.py:
```diff
@@ -89,21 +89,34 @@ def get_latest_version(name: str, allow_insecure: bool) -> T.Tuple[str, str]:
+        try:
+            install_one(name, options.allow_insecure)
+        except WrapException as e:
+            print(str(e))
```
Can you print using mlog.exception here? That will give it a red "ERROR:" prefix, which I think is especially important if it's not going to be the last line of output before the nonzero exit.

eli-schwartz commented on docs/markdown/Using-wraptool.md:
```diff
@@ -64,7 +64,12 @@ the install command.
 
 Now you can issue a `subproject('libjpeg')` in your `meson.build` file
-to use it.
+to use it. Since *1.5.0* you may also specify multiple dependencies to install.
```
Since this missed the previous release(s), through no fault of your own, this version number does need updating again. Sorry. :(

eli-schwartz reviewed (changes requested):
Sorry for not re-reviewing this earlier. I guess it slipped under my radar (multiple times???)

I think I am mostly happy with this, I did notice one other issue based on the changes I asked you to make the last time I looked at this PR. But once that is fixed, this PR lgtm and I think it is ready to merge.

bonzini committed:
wraptool: support installing wraps for multiple projects

Previously, `meson wrap install` took only one argument, so installing
multiple projects needed multiple invocations, which is not too convenient.
So add support for installing multiple projects.
